rgdps/usecases/friend_requests.py

```diff
@@ -0,0 +1,233 @@
+        include_deleted=False,
+    )
+
+    friend_request_responses = []
+    for request in requests:
+        # Swap the sender and recipient according to the from_sender_id flag
+        target_user_id = (
+            request.recipient_user_id if is_sender_user_id else request.sender_user_id
+        )
+        user = await repositories.user.from_id(ctx, target_user_id)
+
+        if user is None:
+            continue
+
+        friend_request_responses.append(
+            FriendRequestResponse(request, user),
+        )
+
+    friend_request_count = (
+        await repositories.friend_requests.get_user_friend_request_count(
+            ctx,
+            user_id,
+            is_sender_user_id=is_sender_user_id,
+        )
+    )
+
+    return PaginatedFriendRequestResponse(
+        friend_request_responses,
+        friend_request_count,
+    )
+
+
+async def mark_as_seen(
+    ctx: Context,
+    user_id: int,
+    request_id: int,
+) -> FriendRequest | ServiceError:
+    request = await repositories.friend_requests.from_id(ctx, request_id)
+
+    if request is None:
+        return ServiceError.FRIEND_REQUEST_NOT_FOUND
+
+    # This is only triggered by recipient.
+    if request.recipient_user_id != user_id:
+        return ServiceError.FRIEND_REQUEST_INVALID_OWNER
+
+    request = await repositories.friend_requests.update_partial(
+        ctx,
+        request_id,
+        seen_ts=datetime.now(),
+    )
+
+    if request is None:
+        return ServiceError.FRIEND_REQUEST_NOT_FOUND
+
+    return request
+
+
+async def accept(
+    ctx: Context,
+    sender_user_id: int,
+    recipient_user_id: int,
+    request_id: int,
+) -> ServiceError | None:
+    request = await repositories.friend_requests.from_id(ctx, request_id)
+
+    if request is None:
+        return ServiceError.FRIEND_REQUEST_NOT_FOUND
+
+    if (
+        request.sender_user_id != sender_user_id
+        or request.recipient_user_id != recipient_user_id
+    ):
+        return ServiceError.FRIEND_REQUEST_INVALID_OWNER
+
+    request = await repositories.friend_requests.update_partial(
+        ctx,
+        request.id,
+        deleted=True,
+    )
+
+    if request is None:
+        return ServiceError.FRIEND_REQUEST_NOT_FOUND
+    
+    if sender_user_id == recipient_user_id:
+        return ServiceError.FRIEND_REQUEST_INVALID_TARGET_ID
+
+    # Create 2 records so it's a mutual friend relationship
+    await repositories.user_relationship.create(
+        ctx,
+        sender_user_id,
+        recipient_user_id,
+        UserRelationshipType.FRIEND,
+    )
+    await repositories.user_relationship.create(
+        ctx,
+        recipient_user_id,
+        sender_user_id,
+        UserRelationshipType.FRIEND,
+    )
+
+
+async def create(
+    ctx: Context,
+    sender_user_id: int,
+    recipient_user_id: int,
+    message: str,
+) -> FriendRequest | ServiceError:
+    if len(message) > 140:
+        return ServiceError.FRIEND_REQUEST_MESSAGE_INVALID_CONTENT
```

Comment:
This sort of validation should be done on the endpoint level.

Reply:
Ok yeah that's wrong, cheers for letting me know.

Reply:
Moved to fastapi validation level in 70a28b7